### PR TITLE
Add Python 3.12 to CI

### PR DESCRIPTION
**Pull request by ianthomas23**
Now that CPython 3.12 has been released, try it out in CI in the hope that it can be included in the upcoming 3.3 release. Also added 3.12 to classifiers.

### CODE REVIEW

**Comment by mattpap** [comment]
So close, only contourpy and matplotlib-base (required by squarify; technically matplotlib is optional, but I'm not sure if conda can handle that).

**Comment by ianthomas23** [comment]
> So close, only contourpy and matplotlib-base (required by squarify; technically matplotlib is optional, but I'm not sure if conda can handle that).

contourpy is fine, I have ensured there are python 3.12 packages on conda-forge already.

**Comment by bryevdv** [comment]
It looks like the packages are working and now there is a pandas deprecation failing things:
> E   DeprecationWarning: datetime.datetime.utcfromtimestamp() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.fromtimestamp(timestamp, datetime.UTC).

**Comment by ianthomas23** [comment]
We will have to hold off on this. The cause of the deprecation warning has been fixed in `python-dateutil` but they need to make a new release on PyPI and conda-forge containing the fix and it seems they are experiencing personnel/permissions issues dateutil/dateutil#1284.

**Comment by graingert** [comment]
@ianthomas23 can you ignore the warning from the dateutil module to get this unblocked?

**Comment by bryevdv** [comment]
@graingert presumably you mean here? 

https://github.com/bokeh/bokeh/blob/273f12589b14f30d7b923e15b80cd9cd9ac96ab0/pyproject.toml#L82-L108

Would that just be a line lile
```
'''ignore:datetime.datetime.utcfromtimestamp() is deprecated and scheduled for removal in a future version:DeprecationWarning''', 
```
?

**Comment by bryevdv** [comment]
@bokeh/core I do think we should wait for this PR to release, it would be a bit weird to have a new major release w/o explicit 3.12 support

**Comment by graingert** [comment]
No like this, note the :dateutil suffix

```
'''ignore:datetime.datetime.utcfromtimestamp() is deprecated and scheduled for removal in a future version:DeprecationWarning:dateutil'', 
```

**Comment by ianthomas23** [comment]
Thanks @graingert, trying it out now...

**Comment by bryevdv** [comment]
Sigh. Unfortunately, the deprecation also happens in `test_defaults` which is run via a different mechanism, so that is still failing. 

But that aside, now there is also 
> ERROR tests/unit/bokeh/core/test_serialization.py - AttributeError: partially initialized module 'pandas' has no attribute '_pandas_datetime_CAPI' (most likely due to a circular import)

I will have to try to repro this conda env and see if I can see this locally.

File: pyproject.toml [hunk 1]
```diff
@@ -105,6 +106,7 @@ filterwarnings = [
     '''ignore:elementwise comparison failed; this will raise an error in the future\.:DeprecationWarning''',
     '''ignore:numba is not installed\. This example will be painfully slow\.:UserWarning''',
     '''ignore:Could not infer format, so each element will be parsed individually, falling back to `dateutil`\. To ensure parsing is consistent and as-expected, please specify a format\.:UserWarning''',
+    '''ignore:datetime.datetime.utcfromtimestamp():DeprecationWarning:dateutil''',
```
Comment (graingert) [inline pyproject.toml:109]:
```suggestion
    '''ignore:datetime\.datetime\.utcfromtimestamp\(\):DeprecationWarning:dateutil''',
```

**Review by graingert — CHANGES_REQUESTED** [review]

**Comment by ianthomas23** [comment]
Now that `pytest` is actually ignoring the warning there are no further `pandas` errors. We are left with one test failure in `test_descriptors.py` related to use of `mock`. I know nothing about `mock` so I am stopping here.

**Comment by graingert** [comment]
That's a pretty easy fix the assert is currently a noop and should be `mock_get.assert_called_once_with((f, "bar", 10), {})`

Or a pytest style assert would be:

```
assert mock_get.mock_calls == [mock.call((f, "bar", 10), {})]
```

**Comment by ianthomas23** [comment]
@graingert If you think it is a pretty easy fix then feel free to try it yourself.

**Comment by graingert** [comment]
@ianthomas23 ok it was a bit more complicated than I originally said, because the noop assertion was also incorrect, and the mock_get and mock_set args were swapped around. I've fixed it with a `mock_calls == [...]` assertion

**Comment by ianthomas23** [comment]
@graingert Thanks for that. I'll close this PR then as it is superseded by #13419.

**Pull request closed without merging**